### PR TITLE
 plan: prepare possible property for logical Sort/TopN/Projection

### PR DESCRIPTION
**Pull request by winoros**
Sort/Top-N/Projection doesn't manage property correctly. e.g. For sort/top-n, if the `ByItems` is the same with group-by items, we actually can choose stream agg but don't in current master.

### CODE REVIEW

**Comment by winoros** [comment]
It just choose hash agg.

File: plan/property_cols_prune.go [hunk 1]
```diff
@@ -40,6 +40,44 @@ func (p *LogicalSelection) preparePossibleProperties() (result [][]*expression.C
 	return p.children[0].preparePossibleProperties()
 }
 
+func (p *LogicalSort) preparePossibleProperties() (result [][]*expression.Column) {
+	childResult := p.children[0].preparePossibleProperties()
+	result = make([][]*expression.Column, 0, 1+len(childResult))
+	result = append(result, childResult...)
```
Comment (lamxTyler) [inline plan/property_cols_prune.go]: Why?

Comment (winoros) [reply plan/property_cols_prune.go]: It can generate nominal sort.

File: plan/property_cols_prune.go [hunk 2]
```diff
@@ -40,6 +40,42 @@ func (p *LogicalSelection) preparePossibleProperties() (result [][]*expression.C
 	return p.children[0].preparePossibleProperties()
 }
 
+func (p *LogicalSort) preparePossibleProperties() (result [][]*expression.Column) {
+	p.children[0].preparePossibleProperties()
+	result = make([][]*expression.Column, 0, 1)
```
Comment (lamxTyler) [inline plan/property_cols_prune.go]: I think we can extract a function for the following.

File: plan/property_cols_prune.go [hunk 3]
```diff
@@ -40,6 +40,42 @@ func (p *LogicalSelection) preparePossibleProperties() (result [][]*expression.C
 	return p.children[0].preparePossibleProperties()
 }
 
+func (p *LogicalSort) preparePossibleProperties() (result [][]*expression.Column) {
+	p.children[0].preparePossibleProperties()
```
Comment (XuHuaiyu) [inline plan/property_cols_prune.go]: Why we need this since there is no variable holding the return value?

Comment (zz-jason) [reply plan/property_cols_prune.go]: @XuHuaiyu We have to recursively call every node's `preparePossibleProperties()` in the plan tree.

File: plan/property_cols_prune.go (hunk 2 above)
Comment (zz-jason) [reply plan/property_cols_prune.go]: +1, maybe name it to "getSortPropertyFromGbyCols" ?

Comment (winoros) [reply plan/property_cols_prune.go]: maybe we can merge this two operator in the future.

Comment (winoros) [reply plan/property_cols_prune.go]: Done

File: plan/property_cols_prune.go [hunk 4]
```diff
@@ -58,9 +116,16 @@ func (p *LogicalJoin) preparePossibleProperties() [][]*expression.Column {
 	} else if p.JoinType == RightOuterJoin {
 		leftProperties = nil
 	}
-	resultProperties := make([][]*expression.Column, len(leftProperties), len(leftProperties)+len(rightProperties))
-	copy(resultProperties, leftProperties)
-	resultProperties = append(resultProperties, rightProperties...)
+	resultProperties := make([][]*expression.Column, len(leftProperties)+len(rightProperties))
```
Comment (zz-jason) [inline plan/property_cols_prune.go]: Why make this change?

Comment (winoros) [reply plan/property_cols_prune.go]: It's a slice of slice. So when do the copy on `resultProperties`, the inner element still shares the same slice. If there's `Projection->Join`, The change of `Projection`'s prop will have influence on `Join`'s prop.

**Comment by XuHuaiyu** [comment]
LGTM

File: plan/property_cols_prune.go (hunk 4 above)
Comment (XuHuaiyu) [reply plan/property_cols_prune.go]: Does there any test case that covered this?

Comment (winoros) [reply plan/property_cols_prune.go]: Without this, two of the existing unit-test will fail.

**Comment by XuHuaiyu** [comment]
ci failed @winoros

**Pull request merged**